apps/module-presale/contracts/Presale.sol

```diff
@@ -0,0 +1,336 @@
+        require(_daiFundingGoal > 0, ERROR_INVALID_DAI_FUNDING_GOAL);
+        require(_percentSupplyOffered > 0, ERROR_INVALID_PERCENT_VALUE);
+        require(_percentSupplyOffered < PPM, ERROR_INVALID_PERCENT_VALUE);
+        require(_beneficiaryAddress != 0x0, ERROR_INVALID_BENEFICIARY_ADDRESS);
+        require(_percentFundingForBenefiriary > 0, ERROR_INVALID_PERCENT_VALUE);
+        require(_percentFundingForBenefiriary < PPM, ERROR_INVALID_PERCENT_VALUE);
+
+        initialized();
+
+        daiToken = _daiToken;
+        _setProjectToken(_projectToken, _projectTokenManager);
+
+        fundraisingPool = _fundraisingPool;
+
+        vestingCliffPeriod = _vestingCliffPeriod;
+        vestingCompletePeriod = _vestingCompletePeriod;
+        fundingPeriod = _fundingPeriod;
+
+        beneficiaryAddress = _beneficiaryAddress;
+        percentFundingForBeneficiary = _percentFundingForBenefiriary;
+
+        daiFundingGoal = _daiFundingGoal;
+        percentSupplyOffered = _percentSupplyOffered;
+
+        _calculateExchangeRate();
+    }
+
+    /*
+     * Public
+     */
+
+    /**
+    * @notice Starts the sale, changing its state to Funding. After the sale is started contributors will be able to purchase project tokens.
+    */
+    function start() public auth(START_ROLE) {
+        require(currentSaleState() == SaleState.Pending, ERROR_INVALID_STATE);
+        startDate = getTimestamp64();
+        emit SaleStarted();
+    }
+
+    /**
+    * @notice Buys project tokens using the provided `_daiToSpend` dai tokens. To calculate how many project tokens will be sold for the provided, dai amount, use daiToProjectTokens(). Each purchase generates a numeric purchaseId (0, 1, 2, etc) for the caller, which can be obtained in the TokensPurchased event emitted, and is required for later refunds.
+    * @param _daiToSpend The amount of dai to spend to obtain project tokens.
+    */
+    function buy(uint256 _daiToSpend) public auth(BUY_ROLE) {
+        require(currentSaleState() == SaleState.Funding, ERROR_INVALID_STATE);
+        require(daiToken.balanceOf(msg.sender) >= _daiToSpend, ERROR_INSUFFICIENT_DAI);
+        require(daiToken.allowance(msg.sender, address(this)) >= _daiToSpend, ERROR_INSUFFICIENT_DAI_ALLOWANCE);
+        require(totalDaiRaised.add(_daiToSpend) <= daiFundingGoal, ERROR_EXCEEDS_FUNDING_GOAL);
+
+        // (buyer) ~~~> dai ~~~> (presale)
+        require(daiToken.transferFrom(msg.sender, address(this), _daiToSpend), ERROR_DAI_TRANSFER_REVERTED);
+
+        // (mint ✨) ~~~> project tokens ~~~> (buyer)
+        uint256 tokensToSell = daiToProjectTokens(_daiToSpend);
+        projectTokenManager.issue(tokensToSell);
+        uint256 purchaseId = projectTokenManager.assignVested(
+            msg.sender,
+            tokensToSell,
+            startDate,
+            startDate.add(vestingCliffPeriod),
+            startDate.add(vestingCompletePeriod),
```

Comment:
Perhaps these are two variables we could save in storage during initialization, to save gas costs for the buyer. Each of these calculations requires 2 `SLOAD`s which makes the math work pretty well in our favour (it would only take ~4 buys for it to be "worth it").